File: flexui-android/flexui/src/main/kotlin/dev/flexui/FlexUI.kt
package dev.flexui

import android.content.Context
import android.view.View
import dev.flexui.actions.ActionDispatcher
import dev.flexui.config.ConfigCache
import dev.flexui.config.ConfigFetcher
import dev.flexui.registry.ComponentRegistry
import dev.flexui.render.FlexRenderer
import dev.flexui.schema.FlexNode
import dev.flexui.schema.FlexParser
import dev.flexui.schema.FlexTheme
import dev.flexui.theme.ThemeDefaults
import java.util.concurrent.Executor
import java.util.concurrent.Executors

/**
 * Main entry point for FlexUI SDK.
 * Provides Java-compatible static API for initializing and rendering FlexUI components.
 */
object FlexUI {
    
    private var isInitialized = false
    private var configFetcher: ConfigFetcher? = null
    private var configCache: ConfigCache? = null
    private var defaultTheme: FlexTheme = ThemeDefaults.getDefaultTheme()
    private var backgroundExecutor: Executor = Executors.newFixedThreadPool(4)
    
    /**
     * Initialize FlexUI with base URL
     */
    @JvmStatic
    fun init(context: Context, baseUrl: String) {
        val config = FlexConfig.Builder(baseUrl).build()
        init(context, config)
    }
    
    /**
     * Initialize FlexUI with configuration
     */
    @JvmStatic
    fun init(context: Context, config: FlexConfig) {
        configFetcher = ConfigFetcher.create(
            config.baseUrl,
            config.connectTimeoutMs,
            config.readTimeoutMs
        )
        configCache = ConfigCache.getInstance(context)
        
        if (config.defaultTheme != null) {
            defaultTheme = ThemeDefaults.mergeWithDefaults(config.defaultTheme)
        }
        
        isInitialized = true
    }
    
    /**
     * Register a custom component factory
     */
    @JvmStatic
    fun registerComponent(type: String, factory: FlexComponentFactory) {
        ComponentRegistry.getInstance().registerComponent(type, factory)
    }
    
    /**
     * Register an action handler
     */
    @JvmStatic
    fun onAction(actionType: String, handler: FlexActionHandler) {
        ActionDispatcher.getInstance().registerHandler(actionType, handler)
    }
    
    /**
     * Register an event handler for callback actions
     */
    @JvmStatic
    fun onEvent(eventName: String, handler: FlexActionHandler) {
        ActionDispatcher.getInstance().registerEventHandler(eventName, handler)
    }
    
    /**
     * Render a screen synchronously
     */
    @JvmStatic
    fun render(
        context: Context, 
        screenId: String, 
        data: Map<String, Any> = emptyMap()
    ): View? {
        ensureInitialized()
        
        return try {
            val configJson = fetchConfigSync(screenId)
            if (configJson != null) {
                renderFromJson(context, configJson, data)
            } else {
                null
            }
        } catch (e: Exception) {
            throw FlexError.renderError("Failed to render screen: $screenId", e)
        }
    }
    
    /**
     * Render a screen asynchronously
     */
    @JvmStatic
    fun renderAsync(
        context: Context,
        screenId: String,
        data: Map<String, Any> = emptyMap(),
        callback: FlexRenderCallback
    ) {
        ensureInitialized()
        
        backgroundExecutor.execute {
            try {
                val view = render(context, screenId, data)
                if (view != null) {
                    // Post back to main thread
                    if (context is android.app.Activity) {
                        context.runOnUiThread {
                            callback.onSuccess(view)
                        }
                    } else {
                        callback.onSuccess(view)
                    }
                } else {
                    val error = FlexError.renderError("No view rendered for screen: $screenId")
                    if (context is android.app.Activity) {
                        context.runOnUiThread {
                            callback.onError(error)
                        }
                    } else {
                        callback.onError(error)
                    }
                }
            } catch (e: Exception) {
                val error = if (e is FlexError) e else FlexError.renderError("Render failed", e)
                if (context is android.app.Activity) {
                    context.runOnUiThread {
                        callback.onError(error)
                    }
                } else {
                    callback.onError(error)
                }
            }
        }
    }
    
    /**
     * Render from JSON string directly
     */
    @JvmStatic
    fun renderFromJson(
        context: Context,
        jsonConfig: String,
        data: Map<String, Any> = emptyMap()
    ): View? {
        return try {
            val config = FlexParser.parseConfig(jsonConfig)
            val theme = ThemeDefaults.mergeWithDefaults(config.theme)
            FlexRenderer.getInstance().render(context, config.root, theme, data)
        } catch (e: Exception) {
            throw FlexError.parseError("Failed to parse JSON config", e)
        }
    }
    
    /**
     * Refresh configuration cache for a screen
     */
    @JvmStatic
    fun refreshConfig(screenId: String) {
        ensureInitialized()
        configCache?.remove(screenId)
    }
    
    /**
     * Clear all cached configurations
     */
    @JvmStatic
    fun clearCache() {
        configCache?.clear()
    }
    
    /**
     * Check if FlexUI is initialized
     */
    @JvmStatic
    fun isInitialized(): Boolean {
        return isInitialized
    }
    
    /**
     * Get current default theme
     */
    @JvmStatic
    fun getDefaultTheme(): FlexTheme {
        return defaultTheme
    }
    
    /**
     * Set default theme
     */
    @JvmStatic
    fun setDefaultTheme(theme: FlexTheme) {
        defaultTheme = ThemeDefaults.mergeWithDefaults(theme)
    }
    
    /**
     * Get registered component types
     */
    @JvmStatic
    fun getRegisteredComponents(): Set<String> {
        return ComponentRegistry.getInstance().getRegisteredTypes()
    }
    
    /**
     * Check if a component type is registered
     */
    @JvmStatic
    fun isComponentRegistered(type: String): Boolean {
        return ComponentRegistry.getInstance().isRegistered(type)
    }
    
    private fun ensureInitialized() {
        if (!isInitialized) {
            throw IllegalStateException("FlexUI not initialized. Call FlexUI.init() first.")
        }
    }
    
    private fun fetchConfigSync(screenId: String): String? {
        val cache = configCache ?: return null
        val fetcher = configFetcher ?: return null
        
        // Try cache first
        val cached = cache.get(screenId)
        if (cached != null) {
            return cached.json
        }
        
        // Fetch from network
        val result = fetcher.fetchConfigSync(screenId)
        if (result.isSuccess()) {
            val json = result.getJsonOrNull()
            if (json != null) {
                cache.put(screenId, json)
                return json
            }
        }
        
        return null
    }
}

File: flexui-ios/Sources/FlexUI/FlexUI.swift
import UIKit
import Foundation

/// Main entry point for FlexUI SDK.
/// Provides a Swift API for initializing and rendering FlexUI components.
public final class FlexUI {
    
    // MARK: - Singleton
    private static let shared = FlexUI()
    
    // MARK: - Private Properties
    private var isInitialized = false
    private var configFetcher: ConfigFetcher?
    private var configCache: ConfigCache?
    private var defaultTheme: FlexTheme = ThemeDefaults.defaultTheme
    private let backgroundQueue = DispatchQueue(label: "flexui.background", qos: .utility)
    
    // MARK: - Configuration
    
    /// Initialize FlexUI with base URL
    public static func configure(baseURL: String) {
        let config = FlexConfig(baseURL: baseURL)
        configure(config)
    }
    
    /// Initialize FlexUI with configuration
    public static func configure(_ config: FlexConfig) {
        shared.configFetcher = ConfigFetcher(
            baseURL: config.baseURL,
            connectTimeout: config.connectTimeout,
            readTimeout: config.readTimeout
        )
        shared.configCache = ConfigCache.shared
        
        if let defaultTheme = config.defaultTheme {
            shared.defaultTheme = ThemeDefaults.merge(theme: defaultTheme)
        }
        
        shared.isInitialized = true
    }
    
    // MARK: - Component Registration
    
    /// Register a custom component factory
    public static func registerComponent(
        _ type: String,
        factory: @escaping FlexComponentFactory
    ) {
        ComponentRegistry.shared.register(type: type, factory: factory)
    }
    
    // MARK: - Action Handling
    
    /// Register an action handler for a specific action type
    public static func onAction(
        _ type: String,
        handler: @escaping FlexActionHandler
    ) {
        ActionDispatcher.shared.registerHandler(type: type, handler: handler)
    }
    
    /// Register an event handler for callback actions
    public static func onEvent(
        _ eventName: String,
        handler: @escaping FlexActionHandler
    ) {
        ActionDispatcher.shared.registerEventHandler(eventName: eventName, handler: handler)
    }
    
    // MARK: - Rendering
    
    /// Render a screen synchronously
    public static func render(
        screenId: String,
        data: [String: Any] = [:]
    ) -> UIView? {
        shared.ensureInitialized()
        
        do {
            guard let configJson = shared.fetchConfigSync(screenId: screenId) else {
                return nil
            }
            return try shared.renderFromJson(configJson, data: data)
        } catch {
            print("FlexUI render error: \\(error)")
            return nil
        }
    }
    
    /// Render a screen asynchronously
    public static func renderAsync(
        screenId: String,
        data: [String: Any] = [:],
        completion: @escaping (Result<UIView, FlexError>) -> Void
    ) {
        shared.ensureInitialized()
        
        shared.backgroundQueue.async {
            do {
                if let view = render(screenId: screenId, data: data) {
                    DispatchQueue.main.async {
                        completion(.success(view))
                    }
                } else {
                    DispatchQueue.main.async {
                        completion(.failure(.renderError("No view rendered for screen: \\(screenId)")))
                    }
                }
            } catch {
                DispatchQueue.main.async {
                    let flexError = error as? FlexError ?? .renderError("Render failed", underlyingError: error)
                    completion(.failure(flexError))
                }
            }
        }
    }
    
    /// Render from JSON string directly
    public static func renderFromJson(
        _ jsonString: String,
        data: [String: Any] = [:]
    ) throws -> UIView? {
        return try shared.renderFromJson(jsonString, data: data)
    }
    
    /// Create a view controller containing the rendered FlexUI view
    public static func viewController(
        screenId: String,
        data: [String: Any] = [:]
    ) -> UIViewController {
        let viewController = FlexViewController()
        
        renderAsync(screenId: screenId, data: data) { result in
            switch result {
            case .success(let view):
                viewController.setFlexView(view)
            case .failure(let error):
                viewController.setError(error)
            }
        }
        
        return viewController
    }
    
    // MARK: - Cache Management
    
    /// Refresh configuration cache for a screen
    public static func refreshConfig(screenId: String) {
        shared.ensureInitialized()
        shared.configCache?.remove(screenId: screenId)
    }
    
    /// Clear all cached configurations
    public static func clearCache() {
        shared.configCache?.clear()
    }
    
    // MARK: - Utility
    
    /// Check if FlexUI is initialized
    public static func isInitialized() -> Bool {
        return shared.isInitialized
    }
    
    /// Get current default theme
    public static func getDefaultTheme() -> FlexTheme {
        return shared.defaultTheme
    }
    
    /// Set default theme
    public static func setDefaultTheme(_ theme: FlexTheme) {
        shared.defaultTheme = ThemeDefaults.merge(theme: theme)
    }
    
    /// Get registered component types
    public static func getRegisteredComponents() -> Set<String> {
        return ComponentRegistry.shared.getRegisteredTypes()
    }
    
    /// Check if a component type is registered
    public static func isComponentRegistered(_ type: String) -> Bool {
        return ComponentRegistry.shared.isRegistered(type: type)
    }
    
    // MARK: - Private Methods
    
    private func ensureInitialized() {
        guard isInitialized else {
            fatalError("FlexUI not initialized. Call FlexUI.configure() first.")
        }
    }
    
    private func fetchConfigSync(screenId: String) -> String? {
        guard let cache = configCache, let fetcher = configFetcher else {
            return nil
        }
        
        // Try cache first
        if let cached = cache.get(screenId: screenId) {
            return cached.json
        }
        
        // Fetch from network
        let result = fetcher.fetchConfigSync(screenId: screenId)
        switch result {
        case .success(let json):
            cache.put(screenId: screenId, json: json)
            return json
        case .failure:
            return nil
        }
    }
    
    private func renderFromJson(
        _ jsonString: String,
        data: [String: Any]
    ) throws -> UIView? {
        let config = try FlexParser.parseConfig(jsonString)
        let theme = ThemeDefaults.merge(theme: config.theme ?? FlexTheme())
        return FlexRenderer.shared.render(node: config.root, theme: theme, data: data)
    }
}

// MARK: - FlexViewController

/// A simple view controller that hosts a FlexUI view
private class FlexViewController: UIViewController {
    private let loadingView = UIActivityIndicatorView(style: .large)
    private let errorLabel = UILabel()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        setupUI()
    }
    
    private func setupUI() {
        view.backgroundColor = .systemBackground
        
        // Setup loading view
        loadingView.translatesAutoresizingMaskIntoConstraints = false
        loadingView.hidesWhenStopped = true
        view.addSubview(loadingView)
        NSLayoutConstraint.activate([
            loadingView.centerXAnchor.constraint(equalTo: view.centerXAnchor),
            loadingView.centerYAnchor.constraint(equalTo: view.centerYAnchor)
        ])
        
        // Setup error label
        errorLabel.translatesAutoresizingMaskIntoConstraints = false
        errorLabel.textColor = .systemRed
        errorLabel.textAlignment = .center
        errorLabel.numberOfLines = 0
        errorLabel.isHidden = true
        view.addSubview(errorLabel)
        NSLayoutConstraint.activate([
            errorLabel.centerXAnchor.constraint(equalTo: view.centerXAnchor),
            errorLabel.centerYAnchor.constraint(equalTo: view.centerYAnchor),
            errorLabel.leadingAnchor.constraint(greaterThanOrEqualTo: view.leadingAnchor, constant: 20),
            errorLabel.trailingAnchor.constraint(lessThanOrEqualTo: view.trailingAnchor, constant: -20)
        ])
        
        loadingView.startAnimating()
    }
    
    func setFlexView(_ flexView: UIView) {
        loadingView.stopAnimating()
        errorLabel.isHidden = true
        
        flexView.translatesAutoresizingMaskIntoConstraints = false
        view.addSubview(flexView)
        NSLayoutConstraint.activate([
            flexView.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor),
            flexView.leadingAnchor.constraint(equalTo: view.leadingAnchor),
            flexView.trailingAnchor.constraint(equalTo: view.trailingAnchor),
            flexView.bottomAnchor.constraint(equalTo: view.bottomAnchor)
        ])
    }
    
    func setError(_ error: FlexError) {
        loadingView.stopAnimating()
        errorLabel.isHidden = false
        errorLabel.text = "Error: \\(error.localizedDescription)"
    }
}

// MARK: - Type Aliases

/// Factory function for creating custom components
public typealias FlexComponentFactory = (FlexProps, FlexTheme) -> UIView

/// Handler function for FlexUI actions
public typealias FlexActionHandler = (FlexAction) -> Void